common/parse.go

```diff
@@ -114,66 +122,175 @@ func (jsonManager) ArrToMap(arr []*ajson.Node) ([]map[string]any, error) {
+		} else {
+			return nil, createKeyNotFoundErr(key)
+		}
+	}
+
+	innerNode, err := node.GetKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if innerNode.IsNull() {
+		return nil, handleNullNode(key, optional)
+	}
+
 	count, err := innerNode.GetNumeric()
 	if err != nil {
-		return 0, ErrNotNumeric
+		return nil, ErrNotNumeric
 	}
 
 	if math.Mod(count, 1.0) != 0 {
-		return 0, ErrNotInteger
+		return nil, ErrNotInteger
 	}
 
-	return int64(count), nil
+	result := int64(count)
+
+	return &result, nil
 }
 
 func (jsonManager) GetArr(node *ajson.Node, key string) ([]*ajson.Node, error) {
+	if !node.HasKey(key) {
+		return nil, createKeyNotFoundErr(key)
+	}
+
 	records, err := node.GetKey(key)
 	if err != nil {
 		return nil, err
 	}
 
 	arr, err := records.GetArray()
 	if err != nil {
-		return nil, ErrNotArray
+		return nil, formatProblematicKeyError(key, ErrNotArray)
 	}
 
 	return arr, nil
 }
 
-func (jsonManager) ArrSize(node *ajson.Node, key string) (int64, error) {
-	innerNode, err := node.GetKey(key)
+func (m jsonManager) ArrSize(node *ajson.Node, keys string) (int64, error) {
+	arr, err := m.GetArr(node, keys)
 	if err != nil {
 		return 0, err
 	}
 
-	if !innerNode.IsArray() {
-		return 0, ErrNotArray
+	return int64(len(arr)), nil
+}
+
+func (m jsonManager) GetStringWithDefault(node *ajson.Node, key string, defaultValue string) (string, error) {
+	result, err := m.GetString(node, key, true)
+	if err != nil {
+		return "", err
+	}
+
+	if result == nil {
+		return defaultValue, err
 	}
 
-	return int64(innerNode.Size()), nil
+	return *result, nil
 }
 
-func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (string, error) {
-	if optional && !node.HasKey(key) {
-		// null value in payload is allowed
-		return "", nil
+func (jsonManager) GetString(node *ajson.Node, key string, optional bool) (*string, error) {
+	if !node.HasKey(key) {
+		if optional {
+			// null value in payload is allowed
+			return nil, nil // nolint:nilnil
+		} else {
+			return nil, createKeyNotFoundErr(key)
+		}
 	}
 
 	innerNode, err := node.GetKey(key)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+
+	if innerNode.IsNull() {
+		return nil, handleNullNode(key, optional)
 	}
 
 	txt, err := innerNode.GetString()
 	if err != nil {
-		return "", ErrNotString
+		return nil, ErrNotString
+	}
+
+	return &txt, nil
+}
+
+// GetNestedNode reaches into the JSON node by zooming via dot separated keys
+// Ex: keys = item.shipping.address => item has object with shipping key and so on.
+func (jsonManager) GetNestedNode(node *ajson.Node, keys []string) (*ajson.Node, error) {
+	var err error
+
+	// traverse nested JSON, use every key to zoom in
+	for _, key := range keys {
+		if !node.HasKey(key) {
+			message := fmt.Sprintf("%v; zoom=%v", key, strings.Join(keys, KeysZoomSeparator))
+
+			return nil, createKeyNotFoundErr(message)
+		}
+
+		node, err = node.GetKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !node.IsObject() {
```

Comment:
This check is undesirable, even if the nested node is a string or another type, it is not a problem and we should return it.

Reply:
I will rename the method to GetNestedObject. That what I meant.
In addition I could introduce a combination of GetNestedObject with [GetInteger, GetString...] so it will be GetNestedString.
The number of methods would drasticly increase. I could make query builder struct. Ex:
```
q := common.JSONQuery(node *ajson.Node, keys ...String)
val, err := q.GetString()
q.GetStringWithDefault()
q.GetObject()
q.GetInteger()
```
This would make querying more explicit. Node, query path via keys and then a finishing method to enforce data type.
What do you think? @laurenzlong

Reply:
Or maybe a `GetNestedNode` to do this generic logic, and then you can have `GetNested<Type>` to unwrap the "zoomed-in" value as the desired type.

Reply:
I haven't considered generics. I have expiremented right now with it and I get into some weird type conversions. Under the hood ajson unwraps float64 converting this to int is easier relying on convential ajson methods. Otherwise would need to have some switch on generic type to determine post processing.
The idea is intriguing to me (especially to use generics in golang as I used it only in Java). I will do the composition style.